cts/CTS_Design_Document.md

```diff
@@ -0,0 +1,341 @@
+
+#### C++
+The C++ backend first sets up the ANARI device. Afterwards, all passed test scenes are initialized and rendered. The pixel data for RGBA and depth channel as well as the ANARI log messages are passed to Python via pybind11.
+
+Example code on how to render scenes can be found in [tests/render/main.cpp](../tests/render/main.cpp):
+```C++
+  auto s = getSceneParameters(scene.c_str());
+
+  auto camera = anari::newObject<anari::Camera>(d, "perspective");
+  anari::setParameter(
+      d, camera, "aspect", (float)g_frameSize.x / (float)g_frameSize.y);
+
+  auto renderer = anari::newObject<anari::Renderer>(d, g_rendererType.c_str());
+  anari::setParameter(d, renderer, "pixelSamples", g_numPixelSamples);
+  anari::setParameter(
+      d, renderer, "backgroundColor", s.getBackgroundColor());
+  anari::commitParameters(d, renderer);
+
+  auto frame = anari::newObject<anari::Frame>(d);
+  anari::setParameter(d, frame, "size", g_frameSize);
+  anari::setParameter(d, frame, "color", ANARI_UFIXED8_RGBA_SRGB);
+
+  anari::setParameter(d, frame, "renderer", renderer);
+  anari::setParameter(d, frame, "camera", camera);
+  anari::setParameter(d, frame, "world", s.getWorld());
+
+  anari::commitParameters(d, frame);
+
+  ...
+
+  anari::render(d, frame);
+  anari::wait(d, frame);
+
+  auto fb = anari::map<uint32_t>(d, frame, "color");
+  auto fb_depth = anari::map<uint32_t>(d, frame, "depth");
+
+  ...
+```
+
+#### Example output
+<p float="left">
+  <img src="./images/random_spheres_0.png" width="300"/>
+  <img src="./images/instanced_cubes_0.png" width="300"/>
+  <img src="./images/instanced_cubes_0_depth.png" width="300"/>
+</p>
+<figcaption>Renderings created with anariRenderTests: random_spheres_0, instanced_cubes_0, instanced_cubes_0 depth channel </figcaption>
+
+### Image comparison “smoke tests”
+
+#### Python API
+
+The Python API performs the image comparison between the previously rendered scenes from an ANARI device and the ground truth. The user can customize various parameters in regards to the comparison. A similar tool is used for the Khronos 3D Commerce Certification Program[^3dc], which can be accessed on github[^3dc-github]. The Python function could look similar to this:
+
+```python
+def compare_images(ground_truth_images, test_images, output_folder = ".", comparison_methods = ["SSIM"], thresholds = None, custom_compare_function = None)
```

Comment:
I have the same comment I made on the call.  That using strings here seems a bit strange.

Reply:
The part that's really unclear is how the comparison methods are used.  Are ALL comparison methods used on each image or is there some matching? Which threshold applies to which comparison method etc.

Suggestion:
Do one comparison method per call to compare_images.  This makes this function much more clear.  If you want to compare multiple ways call this multiple times.  

```python
 def compare_images(ground_truth_images, test_images, comparison_methods = ssim_method, output_folder = "."):
      .....

# if you need to set a custom threshold
my_ssim = lambda image_test, image_base: ssim_method(image_test, image_base, threshold=0.1)
compare_images(ground_truth_images, test_images, comparison_methods = my_ssim)
```

Reply:
I agree, that the current function parameters are a bit confusing.
As stated above, I created the function definitions mainly with a command line tool in mind. Thus, the functions and the function parameters are predefined, and the user selects only the ones which he wants to use. The custom comparison method was added as an option for sophisticated users which want to use the tool as a python module and pass their own function. 
In general, different comparison methods might have multiple different parameters, which should be set statically to some agreed upon values. For example, SSIM has multiple parameters which can be changed such as delta_range channel_axis, gaussian_weights etc. To be able to compare results from different ANARI devices consistently, these parameters should be predefined and not be changeable. The threshold could also be predefined. In this case, we could drop the threshold parameter.

All comparisons are used for each image pair (the ground truth image and the test image). Since this function creates a PDF as result, we need to be able to define multiple comparison methods, otherwise a single PDF which shows the different results for each test scene using multiple metrics is not possible.